### PR TITLE
HB3A Single point measurement peak integration

### LINKED ISSUE #21837: Calculate peak intensity based on one measurement for HB3A interface
For weak peaks, it is not possible to scan a peak.  Instead, one measurement is made on the center of the peak.  An algorithm shall be implemented to use this measurement to calculate and estimate the peak intensity.

### Expected behavior

For such a measurement, integrate peak intensity within region of interest by fit the data to Gaussian.  The intensity of the Gaussian function is the peak height I of the 3D Gaussian peak.  By using the peak width from other measurements with same 2theta value, the peak intensity can be calculated.

After reviewing the new workflow to calculate the single-point-measurement peak intensity, some improvements on UI are required to make.

1. The workspace of scanning SPICE fie in 'Survey' shall be kept in the memory and grouped. These workspaces shall meet the standard naming convention;
2. UI shall load an exported peak integration table;
3. NTable shall check input cell type and store with defined type!

(New) As in the experiment, the single-measurement scan and multiple-measurement scan are mixed.  UI shall automatically separate these two types of scans.

(New) Experiment was run on two different temperatures.  UI will separate scans by temperature too.

### PR DESCRIPTION
**Pull request by wdzhou**
Implement a functionality in HB3A interface to calculate peak intensities of weak peaks which are only measured by single point.

**To test:**

Here is the workflow to calculate single-measurement peak intensity and export to Fullprof data file.

Using Exp 668 as a test case.

* A user specifies the range of scan numbers (289 - 450) that include all the neutron peaks regardless whether they are single-point measurement or multiple-points measurement.
* The user pushes button *Survey*.  Mantid will load all the SPICE files of the scans.
* The user sorts the table by sample temperature ('Sample Temp') to check that all the scans were measured under similar temperature.
* The user finds the multiple-measurements run with highest single-measurement counts and view the raw detector's counts in tab 'View Raw Data'.
* The user define and apply ROI from the strongest multi-measurement nuclear peak.

* The user integrate all multi-measurement scan peaks.  Mantid will generate a report for user to find out whether ROI is well defined or not. Mantid will also generate a report for counts outside of ROI for background.
* The user launches single-point peak integration window from menu *Peak Integraton* and sub menu *Single Pt Integration*. Mantid will add all the single-point scans to the popped out window.
* The user will be prompted with the information about how to map single-point scan to a complete nuclear peak scan with same 2theta.
* The user may load a CSV file to which integrated peak table was exported previously.
* The user pushes button *Retrieve FWHM*. Mantid then will gather FWHM value from integrated peak.
* The user add mapped scans for each single-pt scan to peak processint tab, and integrate peaks.
* The user pushes *Retrieve FHWHM* again to make sure every scan to have an FWHM.
* The user pushes *Integrate* to integrate peaks.
* The user pushes *Export to Table* to add the integrated peaks to peak processing table for final output.

* The user selects *All* single-point-measurement scans by selecting *Single Pt Scans* in combo-box

Fixes #21837. 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/HFIR_4Circle_Reduction/hfctables.py [hunk 1]
```diff
@@ -11,8 +11,10 @@
 import sys
 from HFIR_4Circle_Reduction import fourcircle_utility
 from HFIR_4Circle_Reduction import guiutility
-
+from PyQt4 import QtCore
```
Comment (peterfpeterson) [inline scripts/HFIR_4Circle_Reduction/hfctables.py]: Still has direct dependence on PyQt4

File: scripts/HFIR_4Circle_Reduction/message_dialog.py [hunk 2]
```diff
@@ -7,12 +7,13 @@
...
 from __future__ import (absolute_import, division, print_function)
 from six.moves import range
-from PyQt4 import QtGui, QtCore
+from qtpy.QtWidgets import (QDialog)  # noqa
+
 
 from . import ui_messagebox
 
 
-class MessageDialog(QtGui.QDialog):
```
Comment (peterfpeterson) [inline scripts/HFIR_4Circle_Reduction/message_dialog.py]: Does this need to be a separate class? It looks like a free function kicking off `QMessageBox` would cover it.

File: scripts/HFIR_4Circle_Reduction/mpl2dgraphicsview.py [hunk 3]
```diff
@@ -8,17 +8,15 @@
 from __future__ import (absolute_import, division, print_function)
 import os
 import numpy as np
-
-from PyQt4 import QtGui
-
+from qtpy.QtWidgets import (QWidget, QVBoxLayout, QSizePolicy)
```
Comment (peterfpeterson) [inline scripts/HFIR_4Circle_Reduction/mpl2dgraphicsview.py]: There is an abstraction in the root of `scripts` to make matplotlib qt4/qt5 compatible. Use that.

File: scripts/HFIR_4Circle_Reduction/mplgraphicsview.py [hunk 4]
```diff
@@ -1165,10 +1171,14 @@ def add_plot_1d(self, vec_x, vec_y, y_err=None, color=None, label="", x_label=No
         # Register
         line_key = self._lineIndex
         if plot_error:
-            msg = 'Return from plot is a {0}-tuple: {1} with plot error is {2}\n'.format(len(r), r, plot_error)
-            for i_r in range(len(r)):
-                msg += 'r[%d] = %s\n' % (i_r, str(r[i_r]))
-            raise NotImplementedError(msg)
+            # plot with error bar
+            self._errorBarDict[line_key] = r[0], r[2][0]
+            # data_line = r[0]
```
Comment (peterfpeterson) [inline scripts/HFIR_4Circle_Reduction/mplgraphicsview.py]: Don't add commented out code

File: scripts/HFIR_4Circle_Reduction/mplgraphicsview.py [hunk 5]
```diff
@@ -1362,16 +1377,30 @@ def clear_all_1d_plots(self):
...
-                # error bar
-                plot[0].remove()
-                for line in plot[1]:
-                    line.remove()
-                for line in plot[2]:
-                    line.remove()
-                del self._lineDict[ikey]
+                # error bar: but not likely to be set to _lineDict
+                raise RuntimeError('It is not correct to set a line with error bar to _lineDict')
+                # plot[0].remove()
```
Comment (peterfpeterson) [inline scripts/HFIR_4Circle_Reduction/mplgraphicsview.py]: Don't add commented out code

File: scripts/HFIR_4Circle_Reduction/peak_integration_utility.py [hunk 6]
```diff
@@ -197,8 +332,15 @@ def fit_gaussian_linear_background(vec_x, vec_y, vec_e, start_value_list=None, f
 
     # do second round fit
     assert isinstance(start_value_list, list) and len(start_value_list) == 4, 'Starting value list must have 4 elements'
+
+    #
```
Comment (peterfpeterson) [inline scripts/HFIR_4Circle_Reduction/peak_integration_utility.py]: Move to debug logging or remove the commented out code

File: scripts/HFIR_4Circle_Reduction/plot3dwindow.py [hunk 7]
```diff
@@ -9,15 +9,15 @@
 from six.moves import range
 import sys
 import numpy as np
-from PyQt4 import QtGui, QtCore
+from qtpy.QtWidgets import (QMainWindow)
 
 from . import ui_View3DWidget
```
Comment (peterfpeterson) [inline scripts/HFIR_4Circle_Reduction/plot3dwindow.py]: `load_ui` again

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Lots of little fixes. I haven't gone through everything yet.

File: scripts/HFIR_4Circle_Reduction/guiutility.py [hunk 8]
```diff
@@ -371,102 +374,127 @@ def parse_integers_editors(line_edits, allow_blank=False):
...
+        :param message:
+        :return:
         """
+        self.info_line.setPlainText(message)
+
+        return
+# END-DEF-CLASS
+
 
+# TODO - 20180815 - Clean!
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply scripts/HFIR_4Circle_Reduction/guiutility.py]: Done

File: scripts/HFIR_4Circle_Reduction/hfctables.py (hunk 1 above)
Comment (wdzhou) [reply scripts/HFIR_4Circle_Reduction/hfctables.py]: Corrected

File: scripts/HFIR_4Circle_Reduction/mplgraphicsview.py (hunk 4 above)
Comment (wdzhou) [reply scripts/HFIR_4Circle_Reduction/mplgraphicsview.py]: Done

File: scripts/HFIR_4Circle_Reduction/mplgraphicsview.py (hunk 5 above)
Comment (wdzhou) [reply scripts/HFIR_4Circle_Reduction/mplgraphicsview.py]: Done

**Review by VickieLynch — APPROVED** [review]
Works well using instructions in pull request

**Review by peterfpeterson — APPROVED** [review]
The rest of the Qt4/Qt5 changes will be done in a later PR. This is good to merge.

**Pull request merged**